File: atCoderEnv/problems/tessoku-book/b69/b69.py
from sys import stdin, setrecursionlimit
from typing import List
import math

setrecursionlimit(10**6)


class Edge:
    def __init__(self, to, cap, rev: "Edge"):
        self.to = to
        self.cap = cap
        self.rev = rev


class FordFulkerson:
    def __init__(self, N):
        self.size = N
        self.graph: List[List[Edge]] = [[] for _ in range(N)]

    def add_edge(self, a, b, c):
        e = Edge(b, c, None)
        rev = Edge(a, 0, e)
        e.rev = rev
        self.graph[a].append(e)
        self.graph[b].append(rev)

    def dfs(self, pos, goal, F):
        # ゴールに到着
        if pos == goal:
            return F

        self.visited[pos] = True

        for e in self.graph[pos]:
            # 容量 0 の辺は使えない
            if e.cap == 0:
                continue
            # 既に訪問した頂点に行かない
            if self.visited[e.to]:
                continue
            # 目的地までのパスを探す
            flow = self.dfs(e.to, goal, min(F, e.cap))
            # フローを流せる場合、残余グラフの容量を flow だけ増減させる
            if flow:
                e.cap -= flow
                e.rev.cap += flow
                return flow
        # すべての辺を探索しても見つからなかった
        return 0

    def max_flow(self, s, t):
        ans = 0
        while True:
            self.visited = [False] * self.size
            F = self.dfs(s, t, math.inf)

            # フローを流せなくなったら操作終了
            if F == 0:
                break
            ans += F
        return ans


def main():
    N, M = map(int, stdin.readline().split())
    # 0: start, N+25: goal, 1~N: 従業員, N+1~N+24: 各時間
    graph = FordFulkerson(N+26)

    # スタートから各従業員までを繋ぐ
    for i in range(N):
        graph.add_edge(0, i+1, 10)

    # 各時間とゴールをつなぐ
    for i in range(1, 25):
        graph.add_edge(N+i, N+25, M)

    for i in range(N):
        for j, char in enumerate(list(input())):
            if char == "1":
                graph.add_edge(i+1, N+1+j, 1)

    max_f = graph.max_flow(0, N+25)
    print("Yes" if max_f >= M*24 else "No")


if __name__ == '__main__':
    main()

File: atCoderEnv/problems/tessoku-book/b69/b69.cpp
#include <bits/stdc++.h>
using namespace std;
#include <atcoder/all>
using namespace atcoder;
/* alias */
using ull = unsigned long long;
using ll = long long;
using vi = vector<int>;
using vvi = vector<vi>;  // 二次元vector
using vs = vector<string>;
using pii = pair<int, int>;
/* macro */
#define rep(i, a, n) for (int i = (a); i < (n); ++i)

struct Edge {
  // to: 行き先, cap: 容量, rev: 行き先から見た自分の番号
  int to, cap, rev;
};
class MaximumFlow {
 public:
  // 最初の変数はメンバー初期化リストの記法を用いて設定する
  explicit MaximumFlow(int n) : size(n), used(n + 1, false), G(n + 1) {}

  void add_edge(int a, int b, int c) {
    int current_Ga = G[a].size();  // 現時点でのG[a]の要素数
    int current_Gb = G[b].size();  // 現時点でのG[b]の要素数
    G[a].push_back(Edge({b, c, current_Gb}));
    G[b].push_back(Edge({a, 0, current_Ga}));
  }

  // F はスタートからposに到達する過程での残余グラフの辺の容量の最小値
  int dfs(int pos, int goal, int F) {
    if (pos == goal) return F;
    used[pos] = true;
    // 探索
    for (int i = 0; i < G[pos].size(); i++) {
      // 容量 0 の辺は扱えない
      if (G[pos][i].cap == 0) continue;

      // 既に訪問んした頂点にはいかない
      if (used[G[pos][i].to] == true) continue;

      // 目的地までのパスを探す
      int flow = dfs(G[pos][i].to, goal, min(F, G[pos][i].cap));

      if (flow >= 1) {
        G[pos][i].cap -= flow;
        G[G[pos][i].to][G[pos][i].rev].cap += flow;
        return flow;
      }
    }
    // 全ての辺を探索しても見つからなかった場合
    return 0;
  }

  int max_flow(int start, int goal) {
    int total_flow = 0;
    while (true) {
      for (int i = 0; i <= size; i++) used[i] = false;
      int F = dfs(start, goal, 1000000000);

      // 流せなくなったら終了
      if (F == 0) break;

      total_flow += F;
    }
    return total_flow;
  }

 private:
  int size;
  vector<bool> used;
  vector<vector<Edge>> G;
};

int main() {
  int N, M;
  char C[59][24];
  cin >> N >> M;
  rep(i, 1, N + 1) rep(j, 0, 24) cin >> C[i][j];
  MaximumFlow mf = MaximumFlow(2 + N + 24);
  // 人とstartをつなぐ(働く時間は10時間まで)
  rep(i, 1, N + 1) mf.add_edge(1, 1 + i, 10);
  // 時間とgoalをつなぐ(各時間にM時間必要)
  rep(j, 0, 24) mf.add_edge(1 + N + j + 1, 2 + N + 24, M);

  // 人と時間をつなぐ
  rep(i, 1, N + 1) rep(j, 0, 24) {
    if (C[i][j] == '1') {
      mf.add_edge(1 + i, 1 + N + j + 1, 1);
    }
  }
  if (M * 24 == mf.max_flow(1, 2 + N + 24)) {
    cout << "Yes" << endl;
  } else {
    cout << "No" << endl;
  }
  return 0;
}